### PR TITLE
[Arm64] Add SIMD codegen

### PR DESCRIPTION
**Pull request by sdmaclea**
@dotnet/jit-contrib @dotnet/arm64-contrib PTAL

### CODE REVIEW

File: src/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -3692,6 +3692,996 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+//    simdNode - The GT_SIMD node
+//
+// Return Value:
+//    None.
+//
+// Notes:
+//    Currently, we only recognize SIMDVector<float> and SIMDVector<int>, and
+//    a limited set of methods.
+//
+void CodeGen::genSIMDIntrinsic(GenTreeSIMD* simdNode)
```
Comment (BruceForstall) [inline src/jit/codegenarm64.cpp]: This function looks the same as the xarch one. Should we add a simdcodegencommon.cpp file and put this there (shared by all architectures, maybe with an `#ifdef` or two)?

File: src/jit/codegenarm64.cpp [hunk 2]
```diff
@@ -3692,6 +3692,996 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+            NYI("SIMDIntrinsicSelect lowered during import to (a & sel) | (b & ~sel)");
+            break;
+
+        default:
+            noway_assert(!"Unimplemented SIMD intrinsic.");
+            unreached();
+    }
+}
+
+insOpts genGetSimdInsOpt(bool is16B, var_types elementType)
```
Comment (BruceForstall) [inline src/jit/codegenarm64.cpp]: Please add a standard function header, and make it a member of the CodeGen class.

File: src/jit/codegenarm64.cpp [hunk 3]
```diff
@@ -3692,6 +3692,996 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+            case SIMDIntrinsicBitwiseXor:
+                result = INS_eor;
+                break;
+            case SIMDIntrinsicCast:
+                result = INS_mov;
+                break;
+            case SIMDIntrinsicConvertToDouble:
+            case SIMDIntrinsicConvertToSingle:
+                unreached(); // Not supported for floating source type
+                result = INS_invalid;
```
Comment (BruceForstall) [inline src/jit/codegenarm64.cpp]: this line will probably create an unreachable code warning (apparently these are only errors on our internal builds)

File: src/jit/codegenarm64.cpp [hunk 4]
```diff
@@ -3692,6 +3692,996 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+                result = INS_fmax;
+                break;
+            case SIMDIntrinsicMin:
+                result = INS_fmin;
+                break;
+            case SIMDIntrinsicMul:
+                result = INS_fmul;
+                break;
+            case SIMDIntrinsicNarrow:
+                result = INS_fcvtn; // INS_fcvtn2;
```
Comment (BruceForstall) [inline src/jit/codegenarm64.cpp]: Can you make this comment more explanatory, or remove it?

File: src/jit/codegenarm64.cpp [hunk 5]
```diff
@@ -3692,6 +3692,996 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+    else
+    {
+        getEmitter()->emitIns_R_R_I_I(INS_ins, attr, targetReg, op2Reg, index, 0);
+    }
+
+    genProduceReg(simdNode);
+}
+
+//-----------------------------------------------------------------------------
+// genSIMDIntrinsicUpperSave: save the upper half of a TYP_SIMD32 vector to
```
Comment (BruceForstall) [inline src/jit/codegenarm64.cpp]: Does the ARM64 ABI only save the bottom 16 bytes of 32-byte vectors at calls?

Comment (sdmaclea) [reply src/jit/codegenarm64.cpp]: ARMv8 ARM64 vectors are 16 bytes long.  
For the ABI, only the lower 8 bytes of the callee preserved registers are preserved.

**Comment by sdmaclea** [comment]
I'll fix these issues when I get back to the office.

File: src/jit/codegenarm64.cpp (hunk 1 above)
Comment (sdmaclea) [reply src/jit/codegenarm64.cpp]: I'll do this in a separate PR

File: src/jit/codegenarm64.cpp [hunk 6]
```diff
@@ -3691,6 +3691,981 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+//    simdNode - The GT_SIMD node
+//
+// Return Value:
+//    None.
+//
+// Notes:
+//    Currently, we only recognize SIMDVector<float> and SIMDVector<int>, and
+//    a limited set of methods.
+//
+// TODO-CLEANUP Merge all versions of this function and move to new file simdcodegencommon.cpp.
```
Comment (sdmaclea) [inline src/jit/codegenarm64.cpp]: Added TODO

File: src/jit/codegenarm64.cpp [hunk 7]
```diff
@@ -3691,6 +3691,981 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+            NYI("SIMDIntrinsicSelect lowered during import to (a & sel) | (b & ~sel)");
+            break;
+
+        default:
+            noway_assert(!"Unimplemented SIMD intrinsic.");
+            unreached();
+    }
+}
+
+insOpts CodeGen::genGetSimdInsOpt(bool is16B, var_types elementType)
```
Comment (sdmaclea) [inline src/jit/codegenarm64.cpp]: Made member of class

File: src/jit/codegenarm64.cpp [hunk 8]
```diff
@@ -3691,6 +3691,981 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+            case SIMDIntrinsicMin:
+                result = INS_fmin;
+                break;
+            case SIMDIntrinsicMul:
+                result = INS_fmul;
+                break;
+            case SIMDIntrinsicNarrow:
+                // Use INS_fcvtn lower bytes of result followed by INS_fcvtn2 for upper bytes
+                // Return lower bytes instruction here
+                result = INS_fcvtn;
```
Comment (sdmaclea) [inline src/jit/codegenarm64.cpp]: Fixed comment

File: src/jit/codegenarm64.cpp [hunk 9]
```diff
@@ -3691,6 +3691,981 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+            case SIMDIntrinsicMin:
+                result = isUnsigned ? INS_umin : INS_smin;
+                break;
+            case SIMDIntrinsicMul:
+                result = INS_mul;
+                break;
+            case SIMDIntrinsicNarrow:
+                // Use INS_xtn lower bytes of result followed by INS_xtn2 for upper bytes
+                // Return lower bytes instruction here
+                result = INS_xtn;
```
Comment (sdmaclea) [inline src/jit/codegenarm64.cpp]: Fixed comment

File: src/jit/codegenarm64.cpp [hunk 10]
```diff
@@ -3691,6 +3691,981 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+                break;
+            case SIMDIntrinsicWidenLo:
+                result = isUnsigned ? INS_uxtl : INS_sxtl;
+                break;
+            case SIMDIntrinsicWidenHi:
+                result = isUnsigned ? INS_uxtl2 : INS_sxtl2;
+                break;
+            default:
+                assert(!"Unsupported SIMD intrinsic");
+                unreached();
```
Comment (sdmaclea) [inline src/jit/codegenarm64.cpp]: Made all unreached cases fall into default cases

File: src/jit/codegenarm64.cpp [hunk 11]
```diff
@@ -3691,6 +3691,981 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+    else
+    {
+        getEmitter()->emitIns_R_R_I_I(INS_ins, attr, targetReg, op2Reg, index, 0);
+    }
+
+    genProduceReg(simdNode);
+}
+
+//-----------------------------------------------------------------------------
+// genSIMDIntrinsicUpperSave: save the upper half of a TYP_SIMD16 vector to
```
Comment (sdmaclea) [inline src/jit/codegenarm64.cpp]: Fixed comment

File: src/jit/codegenarm64.cpp [hunk 12]
```diff
@@ -3691,6 +3691,981 @@ void CodeGen::genEmitHelperCall(unsigned helper, int argSize, emitAttr retSize,
...
+    regNumber op1Reg    = genConsumeReg(op1);
+    assert(op1Reg != REG_NA);
+    assert(targetReg != REG_NA);
+    getEmitter()->emitIns_R_R_I_I(INS_mov, EA_8BYTE, targetReg, op1Reg, 0, 1);
+
+    genProduceReg(simdNode);
+}
+
+//-----------------------------------------------------------------------------
+// genSIMDIntrinsicUpperRestore: Restore the upper half of a TYP_SIMD16 vector to
```
Comment (sdmaclea) [inline src/jit/codegenarm64.cpp]: Fixed comment

**Pull request merged**